### ISSUE TITLE
Is it possible this library has quite a few memory leaks?

### ISSUE THREAD

**Issue by quentinfasquel** [mchoe/SwiftSVG#84]
I am sorry if this isn't really a very detailed issue. I happen to have a few leaks when loading a SVGLayer. If necessary I'll do more research on this.

**Comment by buguibu** [comment]
I'm also facing memory leaks problems, check it:
![captura de pantalla 2018-03-08 a las 16 53 04](https://user-images.githubusercontent.com/2544273/37160743-35fd0ef2-22f1-11e8-8ee9-67f576b839f5.png)
![captura de pantalla 2018-03-08 a las 16 52 54](https://user-images.githubusercontent.com/2544273/37160744-361be142-22f1-11e8-8cec-8c8f18094fba.png)

I'm creating the SVGView  in this way:
`welf.svgView = [SVGView.alloc initWithSVGURL:[NSURL URLWithString:level.imagePath]
                                                  completion:^(SVGLayer * _Nonnull layer) {`
and trying to force deallocation without success:
`    deinit {
        nonOptionalLayer.removeFromSuperlayer()
        layer.removeFromSuperlayer()
        #if DEBUG
            print("🐞👾🐞\(self) \(#function) 💸")
        #endif
    }`

**Comment by buguibu** [comment]
Zooming in, it looks like some memory retain due to closures with **supportedAttrbiutes**:
![captura de pantalla 2018-03-08 a las 17 57 05](https://user-images.githubusercontent.com/2544273/37164314-3686e11e-22fa-11e8-97c0-e0cb6c234371.png)

**Comment by buguibu** [comment]
Easily reproducible with iOS app:
1. visit Examples from GitHub
2. navigate back
3. open "Debug memory graph"
![captura de pantalla 2018-03-15 a las 14 17 53](https://user-images.githubusercontent.com/2544273/37465539-af37db70-285b-11e8-96ce-23953e515db3.png)

I'm not Swift expert but looks like closures in SVGParserSupportedElements _public static var allSupportedElements: SVGParserSupportedElements {_ are leaking memory.

**Comment by quentinfasquel** [comment]
In my situation this could be considered critical. Did anyone take a look at solving it or should I dig into it ?

**Comment by buguibu** [comment]
As i said before I tried to fix it by some blind fixings such as weaking self in closures and simple stuff like that but without luck.

**Comment by buguibu** [comment]
Following the backtrace i ended at Fillable.swfit
```
    var fillAttributes: [String : (String) -> Void] {
        return [
            "color": self.fill,
            "fill": self.fill,
            "fill-opacity": self.fillOpacity,
            "fill-rule": self.fillRule,
            "opacity": self.fillOpacity
        ]
    }
```
as the source of this warning:
![captura de pantalla 2018-05-29 a las 16 58 38](https://user-images.githubusercontent.com/2544273/40667797-45be3cdc-6363-11e8-9941-ec7cebe4c622.png)

**Comment by quentinfasquel** [comment]
Hey all,

I have been investigating a bit today and and what seems to be one of the issues comes from `public static var allSupportedElements: SVGParserSupportedElements`, indeed @buguibu.

I have a version on my hand that uses a function instead of the static var, but the issues is not the function itself as static but what it returns. There is a retain cycle that is somehow caused by Swift ([read here](http://sveinhal.github.io/2016/03/16/retain-cycles-function-references/),

```
let returnElement = SVGPath()
returnElement.supportedAttributes = [
    "d": returnElement.parseD,
    "clip-rule": returnElement.clipRule
]
```

**returnElement.parseD** and **returnElement.clipRule** are responsible for a retain cycle.
`returnElement.parseId` is equal to `SVGPath.parseId(returnElement)` and retains `returnElement`,
using `{ [weak returnElement] args in returnElement?.parseD(args) }` avoid that retain-cycle.

---

Also, a useful thing to make sure your objects are actually releases is to have :
```
deinit {
    print(#function, self)
}
```

---

Finally, there are other reasons for the retain cycle, which I am still investigating in.
However I do not understand why this is labelled as "Performance", as memory issue are critical in my opinion. @mchoe I am using this library in production and am facing a few critical memory issues, I sincerely hope this can be resolved.

**Comment by mchoe** [comment]
Fixed in #102 and available starting v2.2.0
